pkg/lib/operatorclient/service.go

```diff
@@ -28,11 +28,22 @@ func (c *Client) DeleteService(namespace, name string, options *metav1.DeleteOpt
 // UpdateService will update the given Service resource.
 func (c *Client) UpdateService(service *v1.Service) (*v1.Service, error) {
 	klog.V(4).Infof("[UPDATE Service]: %s", service.GetName())
-	oldSa, err := c.GetService(service.GetNamespace(), service.GetName())
+	old, err := c.GetService(service.GetNamespace(), service.GetName())
 	if err != nil {
 		return nil, err
 	}
-	patchBytes, err := createPatch(oldSa, service)
+	normalized, err := cloneAndNormalizeObject(old)
+	if err != nil {
+		return nil, fmt.Errorf("failed to normalize existing Service resource for patch: %w", err)
+	}
+	if service.Spec.ClusterIP == old.Spec.ClusterIP {
+		// Support updating to manifests that specify a
+		// ClusterIP when its value is the same as that of the
+		// existing Service.
+		service = service.DeepCopy()
+		service.Spec.ClusterIP = ""
+	}
+	patchBytes, err := createPatch(normalized, service)
```

Comment:
Would there be an issue applying the patch in the case where the new service specifies a `ClusterIP` whereas the old one does not?

Reply:
Yes, that should fail. My thinking was that there are two scenarios where a patch shouldn't touch `ClusterIP`:

- v1 and v2 omit `ClusterIP`
- v1 and v2 both set the same `ClusterIP` value

For the case where v1 omits `ClusterIP` and v2 does not, I'm not convinced that there is a clear automatic decision to be made. We could go ahead and ignore the fact that v2 specifies a certain IP, but it's plausible that for some reason the operator depends on the v2 service having exactly that IP, so I felt that it was best in that case to force a person to step in and decide what to do.

Reply:
I think that makes sense. I feel like the behavior could be somewhat dependent on the service `Type` -- maybe a loadbalanced or external service would expect different upgrade behavior than a cluster-internal one. I think we can keep an eye on this and build out more robust service upgrades upstream (if it arises) or downstream.